### PR TITLE
Task 38: Upload Logo Masjid

### PR DESCRIPTION
**Pull request by xfrends**
## Deskripsi

Pada PR ini, kita ingin membuat satu fitur agar user dapat meng-upload logo masjid.

## Task Links

- #38 

## Checklist

- [x] Form upload logo masjid di halaman edit profile masjid
- [x] Menggunakan cropperjs untuk crop rasio gambar
- [x] Simpan logo di `storage/app/public` dengan dimensi 200 * 200 px
- [x] Simpan ke nama file ke database pada table `settings` dengan key `masjid_image`
- [x] Tampilkan logo yang sudah terupload di halaman profil masjid
- [x] Tampilkan logo yang sudah terupload di halaman edit profil masjid

### Review

- [x] Semua label dan text yang tampil pakai lang, agar memungkinkan multi bahasa
- [x] Pakai `FILESYSTEM_DRIVER=public`
- [x] Setting key diganti dari `masjid_image` menjadi `masjid_logo_path`
- [x] Cek dan hapus jika ada logo existing (untuk mengoptimalkan storage)
- [x] Memindahkan route dari `web.php` ke `api.php`
- [x] Tambahkan `FILESYSTEM_DRIVER=public` di file `.env.example`
- [x] Ganti function `asset()` dengan `Storage::url()` pada response json `Api/MasjidProfileController.php`
- [x] cropper.css dan cropper.js pakai CDN saja
- [x] Aset Cropper JS pakai yang versi **min (compressed)**
- [x] Route `ajax.masjid_profile.image` diubah jadi `api.masjid_profile.image`
- [x] `{{ Setting::get('masjid_name') ?? 'buku masjid'}}` diganti jadi `{{ Setting::get('masjid_name', config('masjid.name')) }}`

### Screenshots

**Profil masjid dengan Logo**

screen_2024-02-28_002

**Edit profil masjid dengan tombol upload logo**

screen_2024-02-28_003

**Form select and crop image untuk logo 200x200px**

screen_2024-02-28_004

### CODE REVIEW

File: app/Http/Controllers/MasjidProfileController.php [hunk 1]
```diff
@@ -39,4 +40,32 @@ public function update(Request $request): RedirectResponse
...
+    {
+        $this->authorize('edit_masjid_profile');
+
+        $validatedPayload = $request->validate([
+            'image' => 'required'
+        ]);
+
+        if (!base64_decode($validatedPayload['image'])) {
+            return response()->json([
+                'message' => 'Image not found'
```
Comment (nafiesl) [inline app/Http/Controllers/MasjidProfileController.php:54]: Pakai lang, supaya kedepannya bisa multi bahasa.
```suggestion
                'message' => __('masjid_profile.image_not_found'),
```
Jangan lupa tambahkan lang di `resources/lang/(en|id)/masjid_profile.php`

File: app/Http/Controllers/MasjidProfileController.php [hunk 2]
```diff
@@ -39,4 +40,32 @@ public function update(Request $request): RedirectResponse
...
+            ]);
+        }
+
+        $image_parts      = explode(";base64,", $request->image);
```
Comment (nafiesl) [inline app/Http/Controllers/MasjidProfileController.php:58]: Beberapa catatan:
1. Nama variable kita pakai camelCase semua
2. Jarak antara syntax dengan tanda sama dengan cukup 1 spasi ` = `
3. Usahakan selalu pakai tanda petik 1 (`'`), tanda petik 2 hanya dipakai kondisi tertentu saja.
4. Setelah proses validasi, jangan pakai `$request` lagi untuk memanggil data request, kita pakai `$validatedPayload`

```suggestion
        $imageParts = explode(';base64,, $validatedPayload['image']);
```

File: app/Http/Controllers/MasjidProfileController.php [hunk 3]
```diff
@@ -39,4 +40,32 @@ public function update(Request $request): RedirectResponse
...
+        $image_base64     = base64_decode($image_parts[1]);
+        $image_name       = uniqid() . '.webp';
+        $image_full_path  = 'public/' . $image_name;
```
Comment (nafiesl) [inline app/Http/Controllers/MasjidProfileController.php:61]: Ini `public/` nya hardcodded, agak sulit kita mau pindah Storage driver.
```suggestion
        $image_full_path  = $image_name;
```
Sebaiknya kita manfaatkan storage driver `public` saja. Jadi di `.env` dan `.env.example` kita set seperti ini:
```
FILESYSTEM_DRIVER=public
```

File: app/Http/Controllers/MasjidProfileController.php [hunk 4]
```diff
@@ -39,4 +40,32 @@ public function update(Request $request): RedirectResponse
...
+
+        Storage::put($image_full_path, $image_base64);
+        Setting::set('masjid_image', $image_name);
```
Comment (nafiesl) [inline app/Http/Controllers/MasjidProfileController.php:64]: Sebelum kita proses ke sini, sebaiknya kita:
1. Sebelum simpan ke storage dan database, lakukan cek dan hapus dulu image logo yang lama.
2. Untuk setting key, kita lebih explisit bahwa yang kita simpan adalah file path logo, misal: `masjid_logo_path`

```suggestion
       // Cek dulu apakah sudah ada image yang sudah ada
       // Jika sudah ada file logo yang lama, kita hapus dulu.
       
        Storage::put($image_full_path, $image_base64);
        Setting::set('masjid_logo_path', $image_name);
```

File: app/Http/Controllers/MasjidProfileController.php [hunk 5]
```diff
@@ -39,4 +40,32 @@ public function update(Request $request): RedirectResponse
...
+
+        return response()->json([
+            'message' => 'Image Uploaded Successfully',
```
Comment (nafiesl) [inline app/Http/Controllers/MasjidProfileController.php:67]: Pakai lang.

```suggestion
            'message' => __('masjid_profile.logo_uploaded'),
```

File: app/Http/Controllers/MasjidProfileController.php [hunk 6]
```diff
@@ -39,4 +40,32 @@ public function update(Request $request): RedirectResponse
...
+            'image' => asset(Setting::get('masjid_image'))
```
Comment (nafiesl) [inline app/Http/Controllers/MasjidProfileController.php:68]: Kita mesti pakai `Storage::url()` karena kita pakai `FILESYSTEM_DRIVER=public`

```suggestion
            'image' => Storage::url(Setting::get('masjid_image')),
```

File: resources/views/masjid_profile/edit.blade.php [hunk 7]
```diff
@@ -12,6 +12,16 @@
...
                     {!! FormField::textarea('masjid_address', ['required' => true, 'value' => old('masjid_address', Setting::get('masjid_address')), 'label' => __('masjid_profile.address')]) !!}
                     {!! FormField::text('masjid_google_maps_link', ['value' => old('masjid_google_maps_link', Setting::get('masjid_google_maps_link')), 'label' => __('masjid_profile.google_maps_link')]) !!}
+                    <div class="form-group">
+                        <label class="form-label" for="masjid_profile_image">Logo Masjid</label>
+                        <div class="custom-file">
+                          <input type="file" class="custom-file-input image" id="masjid_profile_image" name="masjid_profile_image">
+                          <label class="custom-file-label" for="masjid_profile_image">Choose file</label>
+                        </div>
+                        @if (Setting::get('masjid_image'))
+                            <img id="masjid_profile_image_show" class="img-fluid mt-2" src="{{ asset(Setting::get('masjid_image'))}}" alt="{{ Setting::get('masjid_name') ?? 'buku masjid'}}">
```
Comment (nafiesl) [inline resources/views/masjid_profile/edit.blade.php:22]: 1. `asset()` diganti dengan Storage::url(Setting::get('masjid_logo_path))
2. Di setting itu kita bisa tetapkan default valuenya jika value tidak ditemukan `Setting::get('key', 'default value')`.

```suggestion
                            <img id="masjid_profile_image_show" class="img-fluid mt-2" src="{{ Storage::url(Setting::get('masjid_image')) }}" alt="{{ Setting::get('masjid_name', config('masjid.name')) }}">
```

File: resources/views/masjid_profile/edit.blade.php [hunk 8]
```diff
@@ -21,4 +31,110 @@
...
+              </div>
+          </div>
+        </div>
+        <div class="modal-footer">
+          <button type="button" class="btn btn-secondary" data-dismiss="modal">Cancel</button>
+          <button type="button" class="btn btn-primary" id="crop">Crop</button>
+        </div>
+      </div>
+    </div>
+</div>
```
Comment (nafiesl) [inline resources/views/masjid_profile/edit.blade.php:61]: Yang label-label nya pakai lang ya mas.

File: resources/views/masjid_profile/edit.blade.php [hunk 9]
```diff
@@ -21,4 +31,110 @@
...
+@endsection
+
+@push('scripts')
+<script src="https://cdnjs.cloudflare.com/ajax/libs/cropperjs/1.5.6/cropper.js"></script>
+<script>
+var $modal = $('#modal-masjid');
+var image = document.getElementById('image');
+var cropper;
+
+$(document).on("change", "#masjid_profile_image", function(e){
```
Comment (nafiesl) [inline resources/views/masjid_profile/edit.blade.php:75]: Di sini kita ingin ubah logo, bukan profile image.

```suggestion
$(document).on("change", "#masjid_logo_image", function(e){
```

File: resources/views/masjid_profile/edit.blade.php [hunk 10]
```diff
@@ -21,4 +31,110 @@
...
+                dataType: "json",
+                url: "{{ route('masjid_profile.image')}}",
+                headers: {
+                    'X-CSRF-TOKEN': $('meta[name="csrf-token"]').attr('content')
+                },
+                data: {'_token': $('meta[name="_token"]').attr('content'), 'image': base64data},
+                success: function(data){
+                  $modal.modal('hide');
+                  if(!alert(data.message)) {
+                    console.log(data.image);
```
Comment (nafiesl) [inline resources/views/masjid_profile/edit.blade.php:129]: Ini debug code kita hapus mas.
```suggestion
```

File: resources/views/masjid_profile/show.blade.php [hunk 11]
```diff
@@ -11,6 +11,13 @@
         <div class="card">
             <table class="table table-sm card-table">
                 <tbody>
+                    <tr>
+                        <td colspan="2" class="text-center">
+                        @if (Setting::get('masjid_image'))
+                            <img class="img-fluid my-4" src="{{ asset(Setting::get('masjid_image'))}}" alt="{{ Setting::get('masjid_name') ?? 'buku masjid'}}">
```
Comment (nafiesl) [inline resources/views/masjid_profile/show.blade.php:17]:
```suggestion
                            <img class="img-fluid my-4" src="{{ Storage::url(Setting::get('masjid_image')) }}" alt="{{ Setting::get('masjid_name', config('masjid.name')) }}">
```

File: routes/web.php [hunk 12]
```diff
@@ -103,6 +103,7 @@
     Route::get('masjid_profile', [App\Http\Controllers\MasjidProfileController::class, 'show'])->name('masjid_profile.show');
     Route::get('masjid_profile/edit', [App\Http\Controllers\MasjidProfileController::class, 'edit'])->name('masjid_profile.edit');
     Route::patch('masjid_profile', [App\Http\Controllers\MasjidProfileController::class, 'update'])->name('masjid_profile.update');
+    Route::post('masjid_profile/image', [App\Http\Controllers\MasjidProfileController::class, 'image'])->name('masjid_profile.image');
```
Comment (nafiesl) [inline routes/web.php:106]: Karena ini adalah endpoint API, kita pindahkan route ini ke `routes/api.php`.

**Review by nafiesl — CHANGES_REQUESTED** [review]
Mas @xfrends ijin koreksi mas. Beberapa komentar di bawah.

Kemudian, di proyek ini kita mau **mewajibkan** ada automated testingnya mas. sedangkan di PR ini belum ada. Jika mas bisa, tolong ditambahkan testingnya ya. Jika belum bisa, nanti saya yang akan nambahkan.

File: app/Http/Controllers/Api/MasjidProfileController.php [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+        $imageParts = explode(';base64,', $validatedPayload['image']);
+        $imageBase64 = base64_decode($imageParts[1]);
+        $imageName = uniqid() . '.webp';
+
+        Storage::put($imageName, $imageBase64);
+        Setting::set('masjid_logo_path', $imageName);
+
+        return response()->json([
+            'message' => __('masjid_profile.logo_uploaded'),
+            'image' => asset($imageName)
```
Comment (nafiesl) [inline app/Http/Controllers/Api/MasjidProfileController.php:37]: Bagian sini mesti pakai Storage::url agar bisa tampil sesuai URL yang dari storage.
```suggestion
            'image' => Storage::url($imageName),
```

File: public/css/plugins/cropper.css
Comment (nafiesl) [inline public/css/plugins/cropper.css:1]: Karena kita hanya pakai di 1 tempat, file ini kita tidak perlukan, cukup pakai yang dari CDN saja mas.

File: public/js/plugins/cropper.js
Comment (nafiesl) [inline public/js/plugins/cropper.js:1]: Karena kita hanya pakai di 1 tempat, file ini kita tidak perlukan, cukup pakai yang dari CDN saja mas. Lebih hemat ukuran source code juga.

File: resources/views/masjid_profile/edit.blade.php [hunk 14]
```diff
@@ -21,4 +31,125 @@
...
+</div>
 @endsection
+
+@section('styles')
+    {{ Html::style(url('css/plugins/cropper.css')) }}
+@endsection
+
+@push('scripts')
+    {{ Html::script(url('js/plugins/cropper.js')) }}
+    {{ Html::script(url('js/plugins/noty.js')) }}
```
Comment (nafiesl) [inline resources/views/masjid_profile/edit.blade.php:70]: Kita pakai yang dari CDN saja mas, karena ini hanya di pakai di 1 tempat.

```suggestion
@section('styles')
    {{ Html::style(url('https://cdnjs.cloudflare.com/ajax/libs/cropperjs/1.5.6/cropper.min.css')) }}
@endsection

@push('scripts')
    {{ Html::script(url('https://cdnjs.cloudflare.com/ajax/libs/cropperjs/1.5.6/cropper.min.js')) }}
```

File: resources/views/masjid_profile/edit.blade.php [hunk 15]
```diff
@@ -21,4 +31,125 @@
...
+            canvas.toBlob(function(blob) {
+                url = URL.createObjectURL(blob);
+                var reader = new FileReader();
+                reader.readAsDataURL(blob); 
+                reader.onloadend = function() {
+                    var base64data = reader.result;
+                    $.ajax({
+                        type: "POST",
+                        dataType: "json",
+                        url: "{{ route('ajax.masjid_profile.image')}}",
```
Comment (nafiesl) [inline resources/views/masjid_profile/edit.blade.php:122]: URL nya kita pakai `api` saja, nanti di api.php nya di-update.

```suggestion
                        url: "{{ route('api.masjid_profile.image') }}",
```

File: routes/api.php [hunk 16]
```diff
@@ -42,3 +42,11 @@
      */
     Route::apiResource('books', 'BookController')->names('books');
 });
+
+// Authentication with sessions, specifically for AJAX from Blade.
+Route::group(['middleware' => ['auth', 'web'], 'as' => 'ajax.', 'namespace' => 'Ajax'], function () {
```
Comment (nafiesl) [inline routes/api.php:47]: Menyesuaikan yang di atas.

```suggestion
Route::group(['middleware' => ['auth', 'web'], 'as' => 'api.'], function () {
```

File: resources/views/masjid_profile/show.blade.php [hunk 17]
```diff
@@ -11,6 +11,13 @@
         <div class="card">
             <table class="table table-sm card-table">
                 <tbody>
+                    <tr>
+                        <td colspan="2" class="text-center">
+                        @if (Setting::get('masjid_logo_path'))
+                            <img class="img-fluid my-4" src="{{ Storage::url(Setting::get('masjid_logo_path'))}}" alt="{{ Setting::get('masjid_name') ?? 'buku masjid'}}">
```
Comment (nafiesl) [inline resources/views/masjid_profile/show.blade.php:17]: Di fitur setting itu kita bisa kasih default value mas: `Setting::get('key', 'default_value');`. Jadi default valuenya ambil nama masjid dari config saja.

```suggestion
                            <img class="img-fluid my-4" src="{{ Storage::url(Setting::get('masjid_logo_path'))}}" alt="{{ Setting::get('masjid_name', config('masjid.name')) }}">
```

File: resources/views/masjid_profile/edit.blade.php [hunk 18]
```diff
@@ -12,6 +12,16 @@
...
                     {!! FormField::textarea('masjid_address', ['required' => true, 'value' => old('masjid_address', Setting::get('masjid_address')), 'label' => __('masjid_profile.address')]) !!}
                     {!! FormField::text('masjid_google_maps_link', ['value' => old('masjid_google_maps_link', Setting::get('masjid_google_maps_link')), 'label' => __('masjid_profile.google_maps_link')]) !!}
+                    <div class="form-group" id="masjid-logo">
+                        <label class="form-label" for="masjid_logo_image">{{__('masjid_profile.masjid_logo')}}</label>
+                        <div class="custom-file">
+                          <input type="file" class="custom-file-input image" id="masjid_logo_image" name="masjid_logo_image">
+                          <label class="custom-file-label" for="masjid_logo_image">{{__('masjid_profile.choose_file')}}</label>
+                        </div>
+                        @if (Setting::get('masjid_logo_path'))
+                            <img id="masjid_logo_image_show" class="img-fluid mt-2" src="{{ Storage::url(Setting::get('masjid_logo_path'))}}" alt="{{ Setting::get('masjid_name') ?? 'buku masjid'}}">
```
Comment (nafiesl) [inline resources/views/masjid_profile/edit.blade.php:22]: Di fitur setting itu kita bisa kasih default value mas: `Setting::get('key', 'default_value');`. Jadi default valuenya ambil nama masjid dari config saja.

```suggestion
                            <img id="masjid_logo_image_show" class="img-fluid mt-2" src="{{ Storage::url(Setting::get('masjid_logo_path'))}}" alt="{{ Setting::get('masjid_name', config('masjid.name')) }}">
```

File: resources/views/masjid_profile/edit.blade.php [hunk 19]
```diff
@@ -21,4 +31,125 @@
...
+          <h5 class="modal-title" id="modalLabel">{{__('masjid_profile.edit')}}</h5>
+          <button type="button" class="close" data-dismiss="modal" aria-label="Close">
+            <span aria-hidden="true"></span>
+          </button>
+        </div>
+        <div class="modal-body">
+          <div class="img-container">
+              <div class="row justify-content-center text-center">
+                  <div class="col-md-8 justify-content-center text-center">
+                      <img id="image" src="" alt="{{ Setting::get('masjid_name') ?? 'buku masjid'}}">
```
Comment (nafiesl) [inline resources/views/masjid_profile/edit.blade.php:47]: Di fitur setting itu kita bisa kasih default value mas: `Setting::get('key', 'default_value');`. Jadi default valuenya ambil nama masjid dari config saja.
```suggestion
                      <img id="image" src="" alt="{{ Setting::get('masjid_name', config('masjid.name')) }}">
```

**Review by nafiesl — CHANGES_REQUESTED** [review]
Mas @xfrends ijin nambah review mas.

**Review by nafiesl — APPROVED** [review]
Sip, terima kasih bantuan dan kontribusinya mas @xfrend.

Ini kita merge.
